### PR TITLE
System Manager Enters Existence

### PR DESCRIPTION
**Pull request by Gfcello**
# Description

Creation of the initial implementation of the System Manager. Generally the system manager manages the PM, AM, and TM threads and communicates with the LOS interface. It is the highest level manager on the drone and handles fatal failures.

Currently no Task.

## Testing

Currently untested as only framework complete. Tests will be performed on ability to transition states, start threads, handle communication, and handle fatal failures.

## Documentation

https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/2129887250/System+Manager

# Merge Checklist:

- [x] The changes have been well commented, particularly in hard-to-understand areas.

### CODE REVIEW

File: SystemManager/Inc/SM.hpp
Comment (antholuo) [inline SystemManager/Inc/SM.hpp:23]: nit: do we have to start with an underscore? I thought it was generally discouraged.

File: SystemManager/Inc/SM.hpp
Comment (antholuo) [inline SystemManager/Inc/SM.hpp:32]: nit: style / naming convention.

https://uwarg-docs.atlassian.net/wiki/spaces/ZP/pages/1827635203/Style+Guide

File: SystemManager/Inc/SM.hpp
Comment (antholuo) [inline SystemManager/Inc/SM.hpp:40]: was the final decision to use mailq?

**Review by antholuo — COMMENTED** [review]
Also note that your controller RC Link sig comes through a diff channel (not TM, unless we want it to?).

That needs to be decoded as well and will determine arm/disarm/ user/autopliot level/gps/acro/etc.

File: SystemManager/Inc/SM.hpp
Comment (JackGRWD) [reply SystemManager/Inc/SM.hpp:23]: I'm unfamiliar with how enums work, so I was curious about the = -1 at the end.

Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:23]: Enums (enumerations) just assign a number to each of those names in the SM namespace. It defaults to starting at 0 (so BOOT == 0), but you can overwrite the values. I overwrote the FATAL_FAILURE one to be a negative number so it can be easily identified if we need to.

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:32]: Don't waste your time nitpicking this! This is a long way from actual review!

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:40]: Decision is to go with MailQ until we have something working. It is listed as high priority in the "Post MVP Plans" page. Just don't want to block other development when we have something that works well enough in the mean time.

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:32]: Also good to note *what* the nitpick is. Eg. "tab your public:/private:"

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:28]: Lol, I am just spamming TODOs and Ctrl + F.

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:22]: currently I'm just using it for enums and constants, but once there's time yeah

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp:44]: I built all this thread stuff before AM/PM/TM actually had code merged, so wasn't sure. Interfaces would be better if they handle their own init/thread

File: SystemManager/Inc/SM_DataTypes.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM_DataTypes.hpp:23]: These are probably all outdated by now. I will be grabbing from PM and LOS code when it's available.

File: SystemManager/Inc/SM_States.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM_States.hpp:28]: I have been dying not being able to make them virtual. The ONLY reason I haven't is because I am trying to exactly match the syntax of the state system that flew and was tested 2-3 years ago.

After this MVP deadline I will make these virtual and properly override them and remove the bloat.

File: SystemManager/Src/SM_States.cpp
Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:42]: Good point. I will move starting AM to the exit of Boot. That way it only gets halted when we go to fatal fail. 

We ensure the motors won't spin by closing the MailQ between PM and AM. Then SM forces the actuators to 0

Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:42]: that is the only MailQ that AM reads commands from

Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:42]: PIDs would be from a different place

File: SystemManager/Src/SM_States.cpp
Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:15]: Ctrl + h is my friend

File: SystemManager/Inc/SM.hpp
Comment (kozihkode) [inline SystemManager/Inc/SM.hpp]: Not important not but I think this should be renamed from AM_Waypoints to something else to avoid confusion? Since PM is giving AM a vector, not a waypoint

File: SystemManager/Inc/SM.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM.hpp]: Coming back to this. I think SM will call init functions for the manager objects first. Then to make the operate it will run threads that call the operation function of the managers.

File: SystemManager/Inc/SM_DataTypes.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM_DataTypes.hpp:23]: Will be in the common datatypes file Aadi is working on

File: SystemManager/Inc/SM_States.hpp
Comment (Aidan-B) [inline SystemManager/Inc/SM_States.hpp:34]: I recently learned you can also use delete instead of making these private
```suggestion
        BootMode() = delete;
```

File: SystemManager/Src/SM.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM.cpp:24]: This might need to change, depending on what we want to do for flight test :/

File: SystemManager/Src/SM.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM.cpp:57]: I think the mixes are wrong here?

File: SystemManager/Src/SM.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM.cpp:87]: This needs to be calculated each iteration of the thread, right?
```suggestion
    while (true) {
        // Read MessageQ from SM
            xNextWakeTime = xTaskGetTickCount() + AM_THREAD_PERIOD; //or whatever that variable is called
```

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM_States.cpp:109]: Do we need to add a delay so that it arms and doesn't immediately try to pitch back?

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM_States.cpp:166]: What happens if we delete the task in the middle of AM operation. Is that a concern?

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM_States.cpp:235]: Do we need to continually setState, or will it just re-run the same state if we don't?

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [inline SystemManager/Src/SM_States.cpp:120]: Are there any scenarios where we want to go to GroundOp mode from FlightMode? idk what that would look like, but I imagine after a landing we may want this?

**Review by Aidan-B — CHANGES_REQUESTED** [review]
Lots of good things here! I think the comments I added were pretty minimal.

File: AttitudeManager/Src/AM.cpp
Comment (Gfcello) [reply AttitudeManager/Src/AM.cpp:49]: I believe that these are just IDs, so they don't need to be.

File: SystemManager/Inc/SM_StateManager.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM_StateManager.hpp:28]: Not sure how nice static + virtual is. Since each state has to override it anyway, I don't really see the benefit.

File: SystemManager/Inc/SM_States.hpp
Comment (Gfcello) [reply SystemManager/Inc/SM_States.hpp:34]: Compiler disagrrees
image

File: SystemManager/Src/SM.cpp
Comment (Gfcello) [reply SystemManager/Src/SM.cpp:57]: They are, will need to be updated before flying. I think your PR will be fixing that?

File: SystemManager/Src/SM.cpp
Comment (Gfcello) [reply SystemManager/Src/SM.cpp:24]: I know. If we need I can build one for quad too, but I haven't been asked yet.

File: SystemManager/Src/SM.cpp
Comment (Gfcello) [reply SystemManager/Src/SM.cpp:87]: No, based on the example here it seems that the delay function updates it automatically: https://www.freertos.org/vtaskdelayuntil.html

File: SystemManager/Src/SM_States.cpp
Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:109]: I made sure the throttle is set to 0, so am not too worried. It will be a clear sign the drone is armed.

File: SystemManager/Src/SM_States.cpp
Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:166]: Not really a concern. FreeRTOS is just scheduling these and has I believe an idle task that will remove it from the queue. This makes sure that the one resource it can hold (that global flag that isn't in here yet) won't get locked out.

File: SystemManager/Src/SM_States.cpp
Comment (Gfcello) [reply SystemManager/Src/SM_States.cpp:120]: I think after landing that would make sense. Once landing is added to this it should be easy to add.

File: SystemManager/Inc/SM_StateManager.hpp
Comment (Aidan-B) [reply SystemManager/Inc/SM_StateManager.hpp:28]: The benefit is that it requires all children to implement the getInstance() function.

File: SystemManager/Inc/SM_States.hpp
Comment (Aidan-B) [reply SystemManager/Inc/SM_States.hpp:34]: Ah, you are using the default compiler under the hood, so you're right it would need to be defined privately. Alternatively you can use:
```suggestion
        BootMode() = default;
```

I'll let you decide if you prefer `BootMode() {};`

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [reply SystemManager/Src/SM_States.cpp:109]: I guess I imagined the motors would be on in an "armed" state. I think we can leave this for now, and if it becomes an issue we can address this later.

File: SystemManager/Src/SM_States.cpp
Comment (Aidan-B) [reply SystemManager/Src/SM_States.cpp:235]: I understand that, but I guess I was wondering if there is any need to specify that the next state is FatalFailure if we are already in FatalFailure. I dont think it matters though, we can resolve this.

File: SystemManager/Inc/SM_StateManager.hpp
Comment (Aidan-B) [reply SystemManager/Inc/SM_StateManager.hpp:28]: Did a quick google, looks like you're correct and it cant be easily done. Resolving.

**Review by Aidan-B — APPROVED** [review]
lgtm :)

**Review by AadityaChaudhary — APPROVED** [review]
LGTM

File: Common/Inc/CommonDataTypes.hpp
Comment (kozihkode) [inline Common/Inc/CommonDataTypes.hpp:21]: I think this needs to be renamed to LosSFData

Comment (Gfcello) [reply Common/Inc/CommonDataTypes.hpp]: Resolved in 5af498c

**Review by kozihkode — APPROVED** [review]

**Pull request merged**
